### PR TITLE
Added rustyline support for arrow keys + more 

### PR DESCRIPTION
**Pull request by TheRustyPickle**
<h3>What was added</h3>

- Arrow up and down key to cycle history
- Additional support of the left and right arrow keys to control the cursor, CTRL + C, and more. Check at [rustyline](https://github.com/kkawakam/rustyline)

<h3>Changed</h3>

- `rustyline` currently handles all input events instead of the previous `stdin`

<h3>Up for feedback</h3>

- Whether to remove unused functions `spawn_stdinput_handler` and `read_input`
- Change `line_handling.rs` or `line_handler` to something more concise?
- The working directory is being used to save rustyline `history`. Suggestion for a concrete fixed path?
- `rustyline` can potentially support autocompletion similar to a shell (still unsure how). Create another issue for it?

Related issues: #20 #142. Looking forward to hearing your thoughts, @24seconds! Also, let me know if I can improve the pr message in some way so you can understand more easily.

### CODE REVIEW

**Comment by 24seconds** [comment]
wow thank you for the quick action! I will take a look later (Friday night ~ late night KST 🙇 )

**Comment by 24seconds** [comment]
Sorry. Today was busy day. Let me review this until tomorrow mid afternoon 🙇

**Comment by 24seconds** [comment]
[Update]
I was busy today also due to unexpected accident. Let me finish the review until weekend KST.

File: src/line_handling.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    // let rustyline handle the input events instead. Keep the same architecture as before
+    tokio::spawn(async move {
+        loop {
+            // set up what to show at the beginning of the line
+            let readline = rl.readline("> ");
+
+            match readline {
+                Ok(line) => {
+                    // add each line to history so arrow up/down key can work
+                    rl.add_history_entry(line.as_str()).unwrap();
```
Comment (24seconds) [inline src/line_handling.rs:27]: [comment]
using unwrap here might be okay. I guess error would occur highly unlikely.

File: src/line_handling.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                        .send(UserInput {
+                            input: line,
+                            source: InputSource::StandardInput,
+                        })
+                        .await;
+                }
+                // handles the CTRL + C event
+                Err(ReadlineError::Interrupted) => {
+                    rl.save_history("history.txt").unwrap();
+                    std::process::exit(0);
```
Comment (24seconds) [inline src/line_handling.rs:44]: [nit]
Could you import std::process using `use` keyword?

File: src/main.rs [hunk 3]
```diff
@@ -16,6 +16,7 @@ use database as db;
 mod configuration;
 mod error;
 mod ipc;
+mod line_handling;
```
Comment (24seconds) [inline src/main.rs:19]: [IMO]
Maybe you are considering naming. I guess `input handling` or `input handler` would be proper. Or `input line handling` or `input line handler`.

File: src/main.rs [hunk 4]
```diff
@@ -73,7 +74,7 @@ async fn main() -> Result<(), Box<dyn Error>> {
 
             let stdin_tx = user_input_tx.clone();
             // TODO(young): handle tokio::spawn return value nicely so that we can use `?` inside
-            let _stdinput_handle = spawn_stdinput_handler(stdin_tx);
+            let _input_handle = line_handling::line_handler(stdin_tx);
```
Comment (24seconds) [inline src/main.rs:77]: [IMO]
If we naming the package name as `line_handler` for example, then the function name should be `handle` or something more simpler.

**Review by 24seconds — COMMENTED** [review]
Thank you for the work! I left some nit comments.

**Comment by 24seconds** [comment]
> - Whether to remove unused functions `spawn_stdinput_handler` and `read_input`
> - Change `line_handling.rs` or `line_handler` to something more concise?
> - The working directory is being used to save rustyline `history`. Suggestion for a concrete fixed path?
> - `rustyline` can potentially support autocompletion similar to a shell (still unsure how). Create another issue for it?

1. Good. Let's remove it
2. I left comment about naming. In my opinion, it would be nice to include `input` keyword in naming.
3. I suggest in memory history. I also left comment. Please take a look.
4. Oh! yes! It would be nice to work in another issue if it is complicated. If the code is not that long - under 10 lines (?), then you can do it in this pr.

Thank you @TheRustyPickle

File: src/line_handling.rs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+use tokio::sync::mpsc::Sender;
+use tokio::task::JoinHandle;
+
+/// Handle all cli input events with rustyline
+pub fn line_handler(tx: Sender<UserInput>) -> JoinHandle<()> {
+    let mut rl = DefaultEditor::new().expect("Could not initiate editor");
+
+    // load history if available. Will create a new file if it didn't exist at saving time.
+    // ? change file location to somewhere discreet?
+    if rl.load_history("history.txt").is_err() {
```
[the remark this answers is not in the document]
Comment (TheRustyPickle) [reply src/line_handling.rs:13]: If we use `add_history_entry`, it adds the history to the memory. If `save_history` is not used, the history is never saved. So I think we can remove `load_history` and `save_history` and use only `add_history_entry` to keep track of the history of the current session.

File: src/line_handling.rs (hunk 2 above)
Comment (TheRustyPickle) [reply src/line_handling.rs:44]: Done.

File: src/main.rs (hunk 3 above)
Comment (TheRustyPickle) [reply src/main.rs:19]: Opted for `line_handler` as the package name and `handle` as the function. Lmk if it's alright with you.

**Comment by TheRustyPickle** [comment]
Thank you, @24seconds. Just wanted to confirm this for point 1. If I remove `read_input`, I must also remove a test named `test_read_command`. I will be working on the rest of the points that were made.

File: src/line_handling.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (24seconds) [reply src/line_handling.rs:13]: Got it. Then let's do this

File: src/main.rs (hunk 3 above)
Comment (24seconds) [reply src/main.rs:19]: Good. I have no objection.

**Comment by 24seconds** [comment]
> Thank you, @24seconds. Just wanted to confirm this for point 1. If I remove `read_input`, I must also remove a test named `test_read_command`. I will be working on the rest of the points that were made.

Yes, please remove. Btw, can we write a test case for newly introduced code @TheRustyPickle ?

**Comment by TheRustyPickle** [comment]
Yes, sure. Will work on this as well.

**Comment by TheRustyPickle** [comment]
Hello @24seconds, I couldn't figure out how to add tests for the rustyline code part. Also check the [issue here](https://github.com/kkawakam/rustyline/issues/652) about it. Hope the rest of the code is alright. Thanks for your time!

**Comment by 24seconds** [comment]
Got it. Let me take a look today!

File: src/line_handler.rs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                // handles the CTRL + C event
+                Err(ReadlineError::Interrupted) => {
+                    process::exit(0);
+                }
+                Err(ReadlineError::Eof) => {
+                    process::exit(0);
+                }
+                Err(err) => {
+                    println!("Something went wrong. Error: {:?}", err);
+                    process::exit(0);
```
Comment (24seconds) [inline src/line_handler.rs:44]: [comment]
I think this case we should exit with 1 because something went wrong. What do you think?

**Review by 24seconds — APPROVED** [review]
LGTM! Thank you so much for the work! I enjoyed new feature. Thank you!

File: src/line_handler.rs (hunk 6 above)
Comment (TheRustyPickle) [reply src/line_handler.rs:44]: You are right. Will change it.

**Comment by 24seconds** [comment]
Thank you @TheRustyPickle !! Because of your recent contributions `rust-cli-pomodoro` app has improved a lot!

**Comment by 24seconds** [comment]
I will update pakcage version in cargo and write release note on Sunday

**Comment by TheRustyPickle** [comment]
Thank you for being so cooperative and understanding throughout the process, @24seconds! 🙏

**Pull request merged**